### PR TITLE
[Alerting]: Several modifications in alert rules

### PR DESCRIPTION
**Pull request by papagian**
**What this PR does / why we need it**:
- Use common for, annotations and labels properties for grafana and lotex rules
- Add labels to rules
- Fix update ruleGroup API to return 400 Bad Request response when the request contains a UID that does not exist
- Include the namespace Id in GET responses (required by the frontend)
- Check that the request user has at least edit permissions for managing and deleting rules.
- Return different response statuses for namespace failures

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
Requires https://github.com/grafana/alerting-api/pull/50

### CODE REVIEW

File: pkg/services/ngalert/store/alert_rule.go
Comment (kylebrandt) [inline pkg/services/ngalert/store/alert_rule.go:448]: Is there a reason this object would be nil? I wondering if this should error instead.

**Review by owen-d — APPROVED** [review]

**Review by kylebrandt — APPROVED** [review]
LGTM once the api req is merged, one minor comment and nil, vs error.

Thank you!

File: pkg/services/ngalert/store/alert_rule.go
Comment (owen-d) [reply pkg/services/ngalert/store/alert_rule.go:448]: I believe it's using  sharing the `For`, `Annotations`, and `Labels` field from the `ApiRuleNode` rather than redefining the same fields itself.

Comment (papagian) [reply pkg/services/ngalert/store/alert_rule.go:448]: It can be nil when none of the above fields (`For`, `Annotations`, and `Labels`) is set like in [this](https://github.com/grafana/grafana/blob/d0b600e6e41dcb41fca9c73a98a7c12c1de97d33/pkg/services/ngalert/api/test-data/post-rulegroup-42.json#L79) example.

File: pkg/services/ngalert/api/api_ruler.go
Comment (davidmparrott) [inline pkg/services/ngalert/api/api_ruler.go:29]: More specific error handling 🎉

Comment (gotjosh) [reply pkg/services/ngalert/api/api_ruler.go:29]: The same comment applies to other endpoints that use this function throughout this file so I would consider writing a function for it.

File: pkg/services/ngalert/api/api_ruler.go
Comment (gotjosh) [inline pkg/services/ngalert/api/api_ruler.go:142]: For this one, I see whole bunch of others (coming from `GetFolderByUID`):

- `models.ErrFolderAccessDenied` which should be a 401
- The others coming from `toFolderError`, I would classify most of them as 400 with some of them falling into the 500 criteria.

The latter seems a bit long to tackle right now so maybe we just create an issue to tackle later?

File: pkg/services/ngalert/api/api_ruler.go
Comment (gotjosh) [inline pkg/services/ngalert/api/api_ruler.go:205]:
```suggestion
			return response.Error(http.StatusNotFound, "failed to update rule group", err)
```

File: pkg/services/ngalert/store/alert_rule.go
Comment (gotjosh) [inline pkg/services/ngalert/store/alert_rule.go:161]: Is this correct? If we rewrite this error, this will no longer match in an `errors.Is(err, ngmodels.ErrAlertRuleNotFound)`?

**Review by gotjosh — APPROVED** [review]
Overall looks 👌 , I just have a couple of concerns wrt error handling. I don't need to see this again unless you want me to.

File: pkg/services/ngalert/store/alert_rule.go
Comment (papagian) [reply pkg/services/ngalert/store/alert_rule.go:161]: I think it is since I use %w

Comment (gotjosh) [reply pkg/services/ngalert/store/alert_rule.go:161]: would it be more consistent to use errors.Wrap and then?

File: pkg/services/ngalert/api/api_ruler.go
Comment (papagian) [reply pkg/services/ngalert/api/api_ruler.go:29]: `GetFolderByTitle` calls [toFolderError](https://github.com/grafana/grafana/blob/2ca4f3ecae97b216c063206bd3d6ba8baaaa08af/pkg/services/dashboards/folder_service.go#L241) which transforms the dashboard errors to folder ones.
Then the folder API defines [toFolderError](https://github.com/grafana/grafana/blob/2ca4f3ecae97b216c063206bd3d6ba8baaaa08af/pkg/services/dashboards/folder_service.go#L241) which returns the right response according to the error.
So I think that would be enough to export and use this one as well.

File: pkg/services/ngalert/api/api_ruler.go
Comment (papagian) [reply pkg/services/ngalert/api/api_ruler.go:142]: Regarding the 1st point I think we should return 403 (that's what we return [in any other similar case](https://github.com/grafana/grafana/blob/87c3a2b7904d8a3ece112e1189063f2c8db03223/pkg/api/folder.go#L160))
About the 2nd point, as I stated above I think we can use toFolderError:
https://github.com/grafana/grafana/blob/2ca4f3ecae97b216c063206bd3d6ba8baaaa08af/pkg/api/folder.go#L144-L172

File: pkg/services/ngalert/store/alert_rule.go
Comment (papagian) [reply pkg/services/ngalert/store/alert_rule.go:161]: Do you suggest using [github.com/pkg/errors](https://pkg.go.dev/github.com/pkg/errors) instead of [build-in errors](https://golang.org/pkg/errors/)?
I'm fine with that change but given that so far it's mostly used only inside the `ngalert` module I wonder if it's a deliberate preference in grafana over the built-in one.

File: pkg/services/ngalert/store/alert_rule.go
Comment (gotjosh) [inline pkg/services/ngalert/store/alert_rule.go:317]:
```suggestion
// GetNamespaceByTitle is a handler for retrieving a namespace by its title. Alerting rules follow a Grafana folder-like structure which we call namespaces.
```

File: pkg/services/ngalert/store/alert_rule.go
Comment (gotjosh) [inline pkg/services/ngalert/store/alert_rule.go:161]: Let's leave this as-is for now and add it as a goal to remove as soon as we have test coverage across an API level. I've created a follow up for this - as we're also using it in several other parts of the alerting codebase.

File: pkg/services/ngalert/api/api_ruler.go
Comment (gotjosh) [reply pkg/services/ngalert/api/api_ruler.go]: My only blocker is that I don't think this comment is resolved, `models.ErrDashboardIdentifierNotSet` would result in a 500 in this case if I'm reading this correctly.

Comment (papagian) [reply pkg/services/ngalert/api/api_ruler.go]: The updated code is:
https://github.com/grafana/grafana/blob/d1362e763444e56c3c8bffb9dd7305f37d52f801/pkg/services/ngalert/api/api_ruler.go#L27-L30

and `toNamespaceErrorResponse ` handles this:
https://github.com/grafana/grafana/blob/d1362e763444e56c3c8bffb9dd7305f37d52f801/pkg/services/ngalert/api/api_ruler.go#L251-L255

Comment (papagian) [reply pkg/services/ngalert/api/api_ruler.go]: You are right, I will fix.

**Review by gotjosh — APPROVED** [review]

**Pull request merged**
